orb-ui/cone/src/led.rs

```diff
@@ -0,0 +1,248 @@
+        &self.tx
+    }
+}
+
+/// APA102 LEDs
+struct Apa102<S> {
+    spi: S,
+}
+
+/// Driver implementation for the APA102 LED strip.
+impl<S: SpiBus> Apa102<S> {
+    fn spi_rgb_led_update(&mut self, buffer: &[u8]) -> eyre::Result<()> {
+        const ZEROS: [u8; 4] = [0_u8; 4];
+        let size = buffer.len();
+        // The number of clock pulses required is exactly half the total number of LEDs in the string
+        let led_count = size / 4;
+        let ones_len_bytes = led_count / 8 /* 8 bits per byte */ / 2 + 1 /* at least one stop byte */;
+        let ones = vec![0xFF; ones_len_bytes];
+
+        // Start frame: at least 32 zeros
+        self.spi
+            .write(&ZEROS)
+            .map_err(|e| eyre!("err writing: {e:?}"))?;
+
+        // LED data itself
+        self.spi
+            .write(buffer)
+            .map_err(|e| eyre!("err writing: {e:?}"))?;
+
+        // End frame: at least (size / 4) / 2 ones to clock remaining bits
+        self.spi
+            .write(ones.as_slice())
+            .map_err(|e| eyre!("err writing: {e:?}"))?;
+
+        Ok(())
+    }
+
+    fn spi_rgb_led_update_rgb(
+        &mut self,
+        pixels: &[Argb; CONE_LED_COUNT],
+    ) -> eyre::Result<()> {
+        let mut buffer = vec![0; pixels.len() * 4];
+        for (i, pixel) in pixels.iter().enumerate() {
+            let prefix = if let Some(dimming) = pixel.0 {
+                0xE0 | (dimming & 0x1F)
+            } else {
+                0xE0 | 0x1F
+            };
+
+            // APA102 LED strip uses BGR order
+            buffer[i * 4] = prefix;
+            buffer[i * 4 + 1] = pixel.3;
+            buffer[i * 4 + 2] = pixel.2;
+            buffer[i * 4 + 3] = pixel.1;
+        }
+
+        self.spi_rgb_led_update(buffer.as_slice())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ftdi_embedded_hal::eh1::spi::{Error, ErrorKind, ErrorType};
+    use std::cell::RefCell;
+    use std::fmt::{Debug, Formatter};
+
+    // Mock Spi struct
+    struct MockSpi {
+        written: RefCell<Vec<u8>>,
+    }
+
+    enum MockError {}
+
+    impl Debug for MockError {
+        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+            f.write_str("mocked error")
+        }
+    }
+
+    impl Error for MockError {
+        fn kind(&self) -> ErrorKind {
+            ErrorKind::Other
+        }
+    }
+
+    impl ErrorType for MockSpi {
+        type Error = MockError;
+    }
+
+    impl SpiBus for MockSpi {
+        fn read(&mut self, _words: &mut [u8]) -> Result<(), Self::Error> {
+            todo!()
+        }
+
+        fn write(&mut self, words: &[u8]) -> Result<(), Self::Error> {
+            self.written.borrow_mut().extend_from_slice(words);
+            Ok(())
+        }
+
+        fn transfer(
+            &mut self,
+            _read: &mut [u8],
+            _write: &[u8],
+        ) -> Result<(), Self::Error> {
+            todo!()
```

Comment:
use `unimplemetned` for things you intentionally did not implement, use `todo` for things that you intend to do in the future